cmd/HocusPocus.hs

```diff
@@ -48,18 +50,23 @@ commandParser = Opt.hsubparser $
 makeOptions :: Text -> Opt.Parser a -> Opt.ParserInfo a
 makeOptions headerText parser = Opt.info (Opt.helper <*> parser) (Opt.fullDesc <> Opt.header (toS headerText))
 
+data Error
+  = PeerError Peer.Error
+  | RendezvousError Rendezvous.Error
+  deriving (Eq, Show)
+
 -- | Execute 'Command' against a Wormhole Rendezvous server.
 app :: Command -> Rendezvous.Session -> IO ()
 app command session = do
   result <- runExceptT $ do
     print command
-    nameplate <- ExceptT $ Rendezvous.allocate session
-    mailbox <- ExceptT $ Rendezvous.claim session nameplate
+    nameplate <- ExceptT $ first RendezvousError <$> Rendezvous.allocate session
+    mailbox <- ExceptT $ first RendezvousError <$> Rendezvous.claim session nameplate
     liftIO $ Rendezvous.open session mailbox  -- XXX: We should run `close` in the case of exceptions?
-    liftIO $ Rendezvous.add session (Messages.Phase "foo") (Messages.Body "hahaha")
-    message <- liftIO $ Rendezvous.readFromMailbox session
-    print message
-    ExceptT $ Rendezvous.close session (Just mailbox) (Just Messages.Happy)
+    let (Messages.Nameplate n) = nameplate
+    key <- ExceptT $ first PeerError <$> Peer.pakeExchange session (Spake2.makePassword (toS n <> "-potato"))
+    print key
+    ExceptT $ first RendezvousError <$> Rendezvous.close session (Just mailbox) (Just Messages.Happy)
```

Comment:
Okay, cool.  Now instead of putting a "hahaha" message into the mailbox, we try to put in a message that contains the correct SPAKE2 message to demonstrate we are a party who knows the wormhole code.

I don't quite understand why this change introduces `first RendezvousError <$>` throughout.  I have a hunch that it is to improve the error handling behavior below - to make it operate on `Error` instead of the other random stuff that happens to come out of `Rendezvous.allocate`, `Rendezvous.claim`, etc.  Maybe it is also _required_ to give a uniform type to errors everywhere `ExceptT` is used in this function.  It looks like `Peer.pakeExchange` will fail with `Peer.Error` and the `Rendezvous` APIs will fail with `Rendezvous.Error`.  But I haven't learned enough about ExceptT yet to know if this actually matters.  Seems like it might but the only thing done with `err` is `show` and maybe both `Error` types have the `Show typeclass.

Reply:
> Maybe it is also *required* to give a uniform type to errors everywhere ExceptT is used in this function

This is pretty close to correct.

`runExceptT` turns an `ExceptT err monad a` into an `monad (Either err a)`. Here, `monad` is `IO`, so it turns `ExceptT err IO a` into `IO (Either err a)`.

If you look at the signature of monadic bind (`>>=`) for `ExceptT`, you see:

```haskell
(>>=) :: ExceptT e m a -> (a -> ExceptT e m b) -> ExceptT e m b
```  

Or in our case

```haskell
(>>=) :: ExceptT err IO a -> (a -> ExceptT err IO b) -> ExceptT err IO b
```

From this, you can tell that within a `do` block (which is just a nice syntax for a sequence of `>>=` and `>>`), we cannot change the `err` type at all. It's fixed.

So, it _is_ required to be the same. But! It's not just where `ExceptT` is used—it's everywhere within the `do` block.

`ExceptT` is the inverse of `runExceptT`. It turns an `IO (Either err a)` into an `ExceptT err IO a`. Why would we want to do this? So that we can link up a bunch of `IO (Either err a)` actions, automatically extract the `Right`, and stop whenever one of them returns a `Left`. It's a way of combining the `Either` monad with the `IO` monad.

So, we only call `ExceptT` when we're evaluating an `IO (Either err a)`. If the action is `IO (Either someOtherErr a)`, we need to translate `someOtherErr` into `err` and then convert to `ExceptT err IO a`. 

If it's just a regular `IO a` action, we still need to convert to `ExceptT err IO a`, but there's no error to map. `liftIO` does this for us. (It turns an `IO a` into a `MonadIO m => m a`, and `ExceptT err IO` implements `MonadIO` (see https://hackage.haskell.org/package/mtl-2.2.1/docs/Control-Monad-Except.html#g:2).

>  It looks like Peer.pakeExchange will fail with Peer.Error and the Rendezvous APIs will fail with Rendezvous.Error

Yes, that's right. I added a new `Error` type in this module to unify them into one type, precisely so I could chain them together. 

Other approaches would be to have a non-structured error type (i.e. `Text`), or to have a single, flat error type across the package. I don't really know which is better, but what you see is the direction where my tastes lie.

Mucking around with `ghci` can be a good way to get a grip on these things.